### PR TITLE
build: Update codegen

### LINKED ISSUE #14888: Update to more recent version of code-generator
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
Rook CI is using v0.20 of the code-generator, which is quite old. The latest version is v0.32, so it is time to update.
https://github.com/kubernetes/code-generator/tags

**Expected behavior:**
Rook should use recent tools versions

**How to reproduce it (minimal and precise):**

See Rook's version [here](https://github.com/rook/rook/blob/master/build/makelib/golang.mk#L234):
```
export CODE_GENERATOR_VERSION=0.20.0
```

### PR DESCRIPTION
**Pull request by obnoxxx**
**Description of changes:** 

This updates the codegen tooling to the latest version.
It is based on the changes of PR #14894 and meant as a replacement for and continuation of that PR.

Additionally, this PR contains one  commit with temporary API changes to test code generation in the ci. 
This commit should be removed before merging.

**Issue resolved by this Pull Request:**
Resolves #14888 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by obnoxxx** [comment]
converted to draft and added DNM since this contains temporary API changes for debugging that are not meant to get merged.

**Comment by obnoxxx** [comment]
`make codegen`fails in the ci just like it fails locally for me, but at least locally, it does produce the required changes. In the ci, it does not seem to.

**Comment by obnoxxx** [comment]
locally, `make codegen`fails like this:
```console
$ make codegen
...
e_codegen: /Users/obnox/go/src/github.com/rook/rook/.cache/tools/linux_arm64/code-generator-0.31.1/kube_codegen.sh
Generating deepcopy code for 1 targets
Generating client code for 1 targets
F1127 16:57:28.395068   94837 main.go:65] Error: failed making a parser: error(s) in "github.com/rook/rook/pkg/apis/ceph.rook.io/v1":
-: main module (github.com/rook/rook) does not contain package github.com/rook/rook/pkg/apis/ceph.rook.io/v1
make: *** [codegen] Error 255
$
```

and in the CI it fails almost identically.

The local results (in `pkg/apis/ceph.rook.io/v1/zz_generated.deepcopy.go`) are as expected though.

**Comment by BlaineEXE** [comment]
I found this looking around online for similar issues: https://github.com/kubernetes/code-generator/issues/55

> I solved the problem ,by following this [article](https://medium.com/@trstringer/create-kubernetes-controllers-for-core-and-custom-resources-62fc35ad64a3).

Also https://github.com/kubernetes/code-generator/issues/99 is similar

File: build/codegen/codegen.sh [hunk 1]
```diff
@@ -14,29 +14,55 @@
...
-    client,lister,informer \
-    github.com/rook/rook/pkg/client \
-    github.com/rook/rook/pkg/apis \
-    "${GROUP_VERSIONS}" \
-    --output-base "$(dirname "${BASH_SOURCE[0]}")/../../../../.." \
-    --go-header-file "${scriptdir}/boilerplate.go.txt" \
-    --plural-exceptions "CephNFS:CephNFSes" \
+kube::codegen::gen_client \
+    --output-dir "${basedir}" \
+    --output-pkg "${basedir}/${THIS_PKG}"/pkg/client \
```
[the remark this answers is not in the document]
Comment (obnoxxx) [reply build/codegen/codegen.sh]: Thanks for your hints, @BlaineEXE !

I personally find that more variables for repeating strings  usually make shell scripts easier to understand. But I have removed `THIS_PKG` as you  requested.

I also changed `basedir`to point to the root of the git checkout. But codegen still fails as before.

File: build/codegen/codegen.sh [hunk 2]
```diff
@@ -14,29 +14,58 @@
...
-GROUP_VERSIONS="ceph.rook.io:v1"
+##GROUP_VERSIONS="ceph.rook.io:v1"
+
+scriptdir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+
+basedir="$(cd "${scriptdir}/../.."  && pwd)"
+apisdir="${basedir}"/pkg/apis
+clientdir="${basedir}"/pkg/client
+
+boilerplate="${CODE_GENERATOR}/examples/hack/"boilerplate.go.txt
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:25]:
```suggestion
boilerplate="${basedir}"/build/codegen/boilerplate.go.txt
```

File: build/codegen/codegen.sh [hunk 3]
```diff
@@ -14,29 +14,58 @@
...
-    client,lister,informer \
-    github.com/rook/rook/pkg/client \
-    github.com/rook/rook/pkg/apis \
-    "${GROUP_VERSIONS}" \
-    --output-base "$(dirname "${BASH_SOURCE[0]}")/../../../../.." \
-    --go-header-file "${scriptdir}/boilerplate.go.txt" \
-    --plural-exceptions "CephNFS:CephNFSes" \
+kube::codegen::gen_client \
+    --output-dir "${basedir}" \
+    --output-pkg "${clientdir}" \
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:66]: `--output-pkg` is not a directory per-se. It is a package location with gomod semantics. [Docs](https://github.com/kubernetes/code-generator/blob/b15df6411b47bf6e80bfc63947af6b436b2e05c6/kube_codegen.sh#L400C1-L402C60)

```suggestion
    --output-pkg github.com/rook/rook/pkg/client \
```

Do not use a variable for this. The client dir is (a) not a dir and (b) only used once in the script. Redirection to a var for reading just once only makes the script more complex to read.

File: build/codegen/codegen.sh [hunk 4]
```diff
@@ -14,29 +14,58 @@
...
+    --boilerplate "${boilerplate}" \
+    --with-watch \
+    "${apisdir}" \
+
+##    --plural-exceptions "CephNFS:CephNFSes"
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:71]: Don't forget to add this back

File: build/codegen/codegen.sh [hunk 5]
```diff
@@ -14,29 +14,58 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-GROUP_VERSIONS="ceph.rook.io:v1"
+##GROUP_VERSIONS="ceph.rook.io:v1"
+
+scriptdir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+
+basedir="$(cd "${scriptdir}/../.."  && pwd)"
+apisdir="${basedir}"/pkg/apis
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:22]: `apisdir` is particularly hard to read. Change to use snake case typical of bash scripts

```suggestion
base_dir="$(cd "${scriptdir}/../.."  && pwd)"
apis_dir="${basedir}"/pkg/apis
```

File: build/codegen/codegen.sh [hunk 6]
```diff
@@ -14,29 +14,58 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-GROUP_VERSIONS="ceph.rook.io:v1"
+##GROUP_VERSIONS="ceph.rook.io:v1"
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:17]: ?

File: build/codegen/codegen.sh [hunk 7]
```diff
@@ -14,29 +14,58 @@
...
+echo "DEBUG: apisdir: ${apisdir}"
+echo "DEBUG: clientdir: ${clientdir}"
+echo "DEBUG: CODE_GENERATOR: ${CODE_GENERATOR}"
+echo "DEBUG: boilerplate ${boilerplate}"
+echo "DEBUG: kube_codegen: ${kube_codegen}"
+ echo "DEBUG: done"
+# exit 0
+
+
+
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:40]: Nix when done

File: build/codegen/codegen.sh [hunk 8]
```diff
@@ -14,29 +14,58 @@
...
+set -o pipefail
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:41]: Per the codegen example file, please use this pattern and location within the file: 
https://github.com/kubernetes-sigs/container-object-storage-interface/blob/7ddc93baaa3f08c9c8990a17c7b958955d93c044/client/hack/update-codegen.sh#L17C1-L19C16

This will also mean changing the shebang to `#!/usr/bin/env bash` without the `-S` or `-e`

File: build/codegen/codegen.sh [hunk 9]
```diff
@@ -14,29 +14,58 @@
...
+##GROUP_VERSIONS="ceph.rook.io:v1"
+
+scriptdir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+
+basedir="$(cd "${scriptdir}/../.."  && pwd)"
+apisdir="${basedir}"/pkg/apis
+clientdir="${basedir}"/pkg/client
+
+boilerplate="${CODE_GENERATOR}/examples/hack/"boilerplate.go.txt
+kube_codegen="${CODE_GENERATOR}/kube_codegen.sh"
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:26]: Only used once in the script. No need to redirect to var for this.

File: build/codegen/codegen.sh [hunk 10]
```diff
@@ -14,29 +14,58 @@
...
+
 
-scriptdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 
 # CODE GENERATION
 # we run deepcopy and client,lister,informer generations separately so we can use the flag "--plural-exceptions"
 # which is only known by client,lister,informer binary and not the deepcopy binary
+#
+
+
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:54]: Please get rid of double empty lines before final revision.

File: build/makelib/golang.mk [hunk 11]
```diff
@@ -231,7 +231,7 @@ $(YQ):
 	@curl -JL https://github.com/mikefarah/yq/releases/download/$(YQ_VERSION)/yq_$(REAL_HOST_PLATFORM) -o $(YQ)
 	@chmod +x $(YQ)
 
-export CODE_GENERATOR_VERSION=0.20.0
+export CODE_GENERATOR_VERSION=0.31.1
```
Comment (BlaineEXE) [inline build/makelib/golang.mk:234]: Latest currently

```suggestion
export CODE_GENERATOR_VERSION=0.31.3
```

File: build/codegen/codegen.sh [hunk 12]
```diff
@@ -14,29 +14,58 @@
...
-    github.com/rook/rook/pkg/client \
-    github.com/rook/rook/pkg/apis \
-    "${GROUP_VERSIONS}" \
-    --output-base "$(dirname "${BASH_SOURCE[0]}")/../../../../.." \
-    --go-header-file "${scriptdir}/boilerplate.go.txt"
+ kube::codegen::gen_helpers \
+    --boilerplate "${boilerplate}" \
+    "${apisdir}"
+
+##    "${GROUP_VERSIONS}" \
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:61]: nix

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: build/codegen/codegen.sh (hunk 2 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh:25]: @BlaineEXE wrote:
> We should use Rook's boilerplate file, not the sample

Right. I changed it, but our file is called `header.txt`.
And I could have used `${scriptdir}"`but used the same pattern `"${basedir}"/build/codegen` as you suggested.

File: build/codegen/codegen.sh (hunk 8 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: done.

I wanted to do the shebang change anyway, since without it, `make codegen`does not run at all in my local ubuntu tests.

File: build/codegen/codegen.sh (hunk 9 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh:30]: done. 😢

File: build/codegen/codegen.sh (hunk 10 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: done (at least some.

File: build/codegen/codegen.sh (hunk 12 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: gone

File: build/codegen/codegen.sh (hunk 6 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: gone. That was a leftover.

File: build/codegen/codegen.sh (hunk 3 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: done

File: build/codegen/codegen.sh (hunk 4 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: done

**Comment by obnoxxx** [comment]
@BlaineEXE:

I think I have addressed all your requests and with the latest updates, I see this strange situation:

`make codegen` succeeds (exit code `0`), but no diff is produced.

Before, the situation was this:

`make codegen`failed with errors about pkg not existing but it did create correct deepcopy changes.

Before, the codegen run took a long time and produced quite a lot of screen output.
Now, codegen finishes very quickly and produces no screen output at all.

File: build/codegen/codegen.sh [hunk 13]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env -S bash -e
+#!/usr/bin/env  bash
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:1]:
```suggestion
#!/usr/bin/env bash
```

File: build/codegen/codegen.sh [hunk 14]
```diff
@@ -13,30 +13,50 @@
...
-    deepcopy \
-    github.com/rook/rook/pkg/client \
-    github.com/rook/rook/pkg/apis \
-    "${GROUP_VERSIONS}" \
-    --output-base "$(dirname "${BASH_SOURCE[0]}")/../../../../.." \
-    --go-header-file "${scriptdir}/boilerplate.go.txt"
+ kube::codegen::gen_helpers \
+    --boilerplate "${boilerplate}" \
+    "github.com/rook/rook/pkg/apis" \
+
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:53]: empty line

File: build/codegen/codegen.sh [hunk 15]
```diff
@@ -13,30 +13,50 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+set -o pipefail
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:16]: New empty line above

File: build/codegen/codegen.sh [hunk 16]
```diff
@@ -13,30 +13,50 @@
...
+
 
-GROUP_VERSIONS="ceph.rook.io:v1"
 
-scriptdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 
 # CODE GENERATION
 # we run deepcopy and client,lister,informer generations separately so we can use the flag "--plural-exceptions"
 # which is only known by client,lister,informer binary and not the deepcopy binary
+#
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:47]: nix

**Review by BlaineEXE — COMMENTED** [review]
please also squash commits

File: build/codegen/codegen.sh (hunk 16 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: gone

File: build/codegen/codegen.sh (hunk 14 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh:52]: gone

**Comment by obnoxxx** [comment]
@BlaineEXE  wrote:
> please also squash commits

There is really only one actual commit in this PR.
Plus an API change to test code generation.
This test commit is reverted in a last commit.

My plan was to wait until the actual commit is good and then remove the test and revert commit.

File: build/codegen/codegen.sh (hunk 15 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh:17]: added.

File: build/codegen/codegen.sh [hunk 17]
```diff
@@ -14,29 +14,48 @@
...
-bash ${CODE_GENERATOR}/generate-groups.sh \
-    deepcopy \
-    github.com/rook/rook/pkg/client \
-    github.com/rook/rook/pkg/apis \
-    "${GROUP_VERSIONS}" \
-    --output-base "$(dirname "${BASH_SOURCE[0]}")/../../../../.." \
-    --go-header-file "${scriptdir}/boilerplate.go.txt"
+ kube::codegen::gen_helpers \
+    --boilerplate "${boilerplate}" \
+    "github.com/rook/rook/pkg/apis" \
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:52]: Last line shouldn't have escape

```suggestion
    "github.com/rook/rook/pkg/apis"
```

File: build/codegen/codegen.sh [hunk 18]
```diff
@@ -14,29 +14,48 @@
...
-    "${GROUP_VERSIONS}" \
-    --output-base "$(dirname "${BASH_SOURCE[0]}")/../../../../.." \
-    --go-header-file "${scriptdir}/boilerplate.go.txt" \
+kube::codegen::gen_client \
+    --output-dir "${base_dir}" \
+    --output-pkg "github.com/rook/rook/pkg/client" \
+    --boilerplate "${boilerplate}" \
     --plural-exceptions "CephNFS:CephNFSes" \
+    --with-watch \
+    "github.com/rook/rook/pkg/apis" \
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:61]: Last line shouldn't ave escape

```suggestion
    "github.com/rook/rook/pkg/apis"
```

Comment (BlaineEXE) [inline build/codegen/codegen.sh:61]: This is a directory location, not a package location.

```suggestion
    "${base_dir}/pkg/apis"
```

File: build/codegen/codegen.sh (hunk 17 above)
Comment (BlaineEXE) [inline build/codegen/codegen.sh:52]: Directory location, not package location

```suggestion
    "${base_dir}/pkg/apis" \
```

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
I'm poking around my local checkout of this.

```
+ /Users/blaine/go/bin/client-gen -v 0 --go-header-file /Users/blaine/go/src/github.com/rook/rook/build/codegen/header.txt --output-dir /Users/blaine/go/src/github.com/rook/rook/clientset --output-pkg github.com/rook/rook/pkg/client/clientset --clientset-name versioned --apply-configuration-package '' --input-base /Users/blaine/go/src/github.com/rook/rook/pkg/apis --plural-exceptions CephNFS:CephNFSes --input ceph.rook.io/v1
F1202 14:26:09.518754   19700 main.go:65] Error: failed making a parser: error(s) in "github.com/rook/rook/pkg/apis/ceph.rook.io/v1":
-: main module (github.com/rook/rook) does not contain package github.com/rook/rook/pkg/apis/ceph.rook.io/v1
```

I've realized that the error is complaining that `github.com/rook/rook/pkg/apis` is a separate go module (a submodule) from the main module (github.com/rook/rook). 

Rook's API is nothing but the API.
The generated client code is part of the main module.

COSI treats the API and the generated client code all as the same go submodule, so it doesn't end up having the same issue.

File: build/codegen/codegen.sh [hunk 19]
```diff
@@ -14,29 +14,48 @@
...
-    "${GROUP_VERSIONS}" \
-    --output-base "$(dirname "${BASH_SOURCE[0]}")/../../../../.." \
-    --go-header-file "${scriptdir}/boilerplate.go.txt" \
+kube::codegen::gen_client \
+    --output-dir "${base_dir}" \
+    --output-pkg "github.com/rook/rook/pkg/client" \
+    --boilerplate "${boilerplate}" \
     --plural-exceptions "CephNFS:CephNFSes" \
+    --with-watch \
+    "${base_dir}/pkg/apis"
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:61]: I got it working. It seems to be that the codegen needs to be run from within the API go module. Updating the script to `cd` there in a subshell will work well

```sh
(
    # because apis is separate gomod, must change to that dir for generation
    cd ${base_dir}/pkg/apis

    # run code deepcopy generation
    kube::codegen::gen_helpers \
        --boilerplate "${boilerplate}" \
        "${base_dir}/pkg/apis"

    # run code client,lister,informer generation
    kube::codegen::gen_client \
        --output-dir "${base_dir}/pkg/client" \
        --output-pkg "github.com/rook/rook/pkg/client" \
        --boilerplate "${boilerplate}" \
        --plural-exceptions "CephNFS:CephNFSes" \
        --with-watch \
        "${base_dir}/pkg/apis"
)
```

Comment (obnoxxx) [reply build/codegen/codegen.sh]: Fantastic, this works!

Thanks a lot, @BlaineEXE !
I updated the code and added you as co-author to the commit.

File: build/codegen/codegen.sh (hunk 18 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: applied

File: build/codegen/codegen.sh (hunk 17 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: applied

**Comment by obnoxxx** [comment]
Note for reviewers:

This PR has 4 commits of which only one would remain in the final form:
1. the actual code change.
2. update of generated files. This might be necessary but see the discussion in a different PR: https://github.com/rook/rook/pull/15075#discussion_r1866321849 
3. temporary API change for testing codegen and its revert: both commits should be removed and not get merged.

File: build/codegen/codegen.sh [hunk 20]
```diff
@@ -14,29 +14,61 @@
...
 # we run deepcopy and client,lister,informer generations separately so we can use the flag "--plural-exceptions"
 # which is only known by client,lister,informer binary and not the deepcopy binary
 
+
+
+(
+# because apis is a separate go module, code generation
+# must be run from within the corresponding directory.
+
+pushd ${base_dir}/pkg/apis
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:55]: `pushd` and `popd` are not always script safe in the face of errors. Please use `( cd ; ... )` as I requested in my comment yesterday. Please also indent the script contents inside the `( )` subshell.

File: build/codegen/codegen.sh [hunk 21]
```diff
@@ -14,29 +14,61 @@
...
-GROUP_VERSIONS="ceph.rook.io:v1"
+set -o pipefail
+set -o nounset
+set -o errexit
+
+
+
+script_dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+
+base_dir="$(cd "${script_dir}/../.."  && pwd)"
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:25]:
```suggestion
script_dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
base_dir="$(cd "${script_dir}/../.."  && pwd)"
```

File: build/codegen/codegen.sh [hunk 22]
```diff
@@ -14,29 +14,61 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-GROUP_VERSIONS="ceph.rook.io:v1"
+set -o pipefail
+set -o nounset
+set -o errexit
+
+
+
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:22]:
```suggestion

```

File: build/codegen/codegen.sh [hunk 23]
```diff
@@ -14,29 +14,61 @@
...
+# echo "DEBUG: script_dir: ${script_dir}"
+# echo "DEBUG: CODE_GENERATOR: ${CODE_GENERATOR}"
+# echo "DEBUG: boilerplate ${boilerplate}"
+#  echo "DEBUG: done"
+# exit 0
+
+
+
+
+
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:39]:
```suggestion

```

File: build/codegen/codegen.sh [hunk 24]
```diff
@@ -14,29 +14,61 @@
...
+source "${CODE_GENERATOR}/kube_codegen.sh"
+
+
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:43]:
```suggestion

```

File: build/codegen/codegen.sh [hunk 25]
```diff
@@ -14,29 +14,61 @@
...
 
-scriptdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 
 # CODE GENERATION
 # we run deepcopy and client,lister,informer generations separately so we can use the flag "--plural-exceptions"
 # which is only known by client,lister,informer binary and not the deepcopy binary
 
+
+
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:50]:
```suggestion
```

File: build/codegen/codegen.sh (hunk 20 above)
Comment (BlaineEXE) [inline build/codegen/codegen.sh:55]:
```suggestion
(
  # because apis is a separate go module, code generation
  # must be run from within the corresponding directory.
  cd ${base_dir}/pkg/apis
```

File: build/codegen/codegen.sh [hunk 26]
```diff
@@ -14,29 +14,61 @@
...
+(
+# because apis is a separate go module, code generation
+# must be run from within the corresponding directory.
+
+pushd ${base_dir}/pkg/apis
+
+
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:57]:
```suggestion

```

File: build/codegen/codegen.sh [hunk 27]
```diff
@@ -14,29 +14,61 @@
...
-    --go-header-file "${scriptdir}/boilerplate.go.txt" \
+kube::codegen::gen_client \
+    --output-dir "${base_dir}/pkg/client" \
+    --output-pkg "github.com/rook/rook/pkg/client" \
+    --boilerplate "${boilerplate}" \
     --plural-exceptions "CephNFS:CephNFSes" \
+    --with-watch \
+    "${base_dir}/pkg/apis"
+
+popd
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:72]:
```suggestion
```

File: build/codegen/codegen.sh [hunk 28]
```diff
@@ -14,29 +14,61 @@
...
+)
+
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:74]:
```suggestion
```

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: build/codegen/codegen.sh (hunk 22 above)
Comment (BlaineEXE) [inline build/codegen/codegen.sh:22]:
```suggestion

```

Comment (obnoxxx) [reply build/codegen/codegen.sh]: applied

File: build/codegen/codegen.sh (hunk 20 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: ood point! I did not get the point with the subshell first. Changed accordingly now.

Comment (obnoxxx) [reply build/codegen/codegen.sh]: applied, conceptually.

File: build/codegen/codegen.sh (hunk 27 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: applied, conceptually.

File: build/codegen/codegen.sh (hunk 25 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: applied

File: build/codegen/codegen.sh (hunk 21 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: applied

File: build/codegen/codegen.sh [hunk 29]
```diff
@@ -14,29 +14,37 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-GROUP_VERSIONS="ceph.rook.io:v1"
+set -o pipefail
+set -o nounset
+set -o errexit
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:19]: I requested that these be applied exactly as in the example update codegen script, but the ordering here is reversed. errexit, nounset, pipefail is the standard ordering I see across many scripts. It's also in alphabetical order. There are also maintenance benefits to keeping our script boilerplate components identical to the one its based on.

I want to note that I take great care when I do code reviews to be clear and exact with my feedback. I consistently see that what I request gets modified in some way in the next iteration without seeming reason or purpose. I'm not sure if this is a misunderstanding that it seems like what I'm asking is unspecific, or if it's deliberately contrary. I want to make it clear that more often than not, I am being specific about what change I want to see and not general. When I give generalized feedback, I try to also make sure that is clear.
 
```suggestion
set -o errexit
set -o nounset
set -o pipefail
```

File: build/codegen/codegen.sh [hunk 30]
```diff
@@ -14,29 +14,37 @@
...
+script_dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:20]: Script error setup is not related to determining the script's dir. Separate these lines

```suggestion
set -o errexit

script_dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
```

File: build/codegen/codegen.sh [hunk 31]
```diff
@@ -14,29 +14,37 @@
...
 
-scriptdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+base_dir="$(cd "${script_dir}/../.."  && pwd)"
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:22]: Finding the repo base dir is IMO related to finding the script dir, so keep `script_dir` and `base_dir` lines together.

File: build/codegen/codegen.sh [hunk 32]
```diff
@@ -14,29 +14,37 @@
...
+
+boilerplate="${base_dir}"/build/codegen/header.txt
+source "${CODE_GENERATOR}/kube_codegen.sh"
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:25]: recording the location of the boilerplate header file is not related to loading the codegen file. Separate these lines.

```suggestion
boilerplate="${base_dir}"/build/codegen/header.txt

source "${CODE_GENERATOR}/kube_codegen.sh"
```

File: build/codegen/codegen.sh [hunk 33]
```diff
@@ -14,29 +14,37 @@
...
 
 # CODE GENERATION
 # we run deepcopy and client,lister,informer generations separately so we can use the flag "--plural-exceptions"
 # which is only known by client,lister,informer binary and not the deepcopy binary
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:29]: This comment doesn't apply to the latest tooling, which doesn't use separate generators. Without it, it's just `# CODE GENERATION` which doesn't add value, so let's just nix this comment altogether

File: build/codegen/codegen.sh [hunk 34]
```diff
@@ -14,29 +14,37 @@
...
 
+(
+# because apis is a separate go module, code generation
+# must be run from within the corresponding directory.
+
+cd  ${base_dir}/pkg/apis
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:35]: The comment directly relates to the directory change, so these should touch

This is inside a subshell, so this should be indented

```suggestion
  # because apis is a separate go module, code generation
  # must be run from within the corresponding directory.
  cd  ${base_dir}/pkg/apis
```

File: build/codegen/codegen.sh [hunk 35]
```diff
@@ -14,29 +14,37 @@
...
-    --output-base "$(dirname "${BASH_SOURCE[0]}")/../../../../.." \
-    --go-header-file "${scriptdir}/boilerplate.go.txt" \
+kube::codegen::gen_client \
+    --output-dir "${base_dir}/pkg/client" \
+    --output-pkg "github.com/rook/rook/pkg/client" \
+    --boilerplate "${boilerplate}" \
     --plural-exceptions "CephNFS:CephNFSes" \
+    --with-watch \
+    "${base_dir}/pkg/apis"
+
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:49]: Empty line here unnecessary

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
Please squash all the commits into a single one when ready to merge. Leaving the temporary commit and its reversion in the git tree only adds clutter. Updating a script and the files it generates separately is not a standard convention. In general, we always want to see one single commit per pull request in Rook's repo. Even if this is different from other projects, that is the standard for this project.

File: build/codegen/codegen.sh (hunk 31 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh:22]: applied

File: build/codegen/codegen.sh (hunk 33 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: done

File: build/codegen/codegen.sh (hunk 35 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: gone

**Comment by obnoxxx** [comment]
@BlaineEXE : all commits squashed as per you request

File: build/codegen/codegen.sh [hunk 36]
```diff
@@ -14,29 +14,34 @@
...
-# CODE GENERATION
-# we run deepcopy and client,lister,informer generations separately so we can use the flag "--plural-exceptions"
-# which is only known by client,lister,informer binary and not the deepcopy binary
+base_dir="$(cd "${script_dir}/../.."  && pwd)"
 
+boilerplate="${base_dir}"/build/codegen/header.txt
+
+source "${CODE_GENERATOR}/kube_codegen.sh"
+
+
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:29]:
```suggestion

```

File: build/codegen/codegen.sh [hunk 37]
```diff
@@ -14,29 +14,34 @@
...
+(
+  # because apis is a separate go module, code generation
+  # must be run from within the corresponding directory.
+  cd  ${base_dir}/pkg/apis
 # run code deepcopy generation
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:34]: The deepcopy comment doesn't relate to the `cd`, so these lines should be separated.

File: build/codegen/codegen.sh [hunk 38]
```diff
@@ -14,29 +14,34 @@
...
-    "${GROUP_VERSIONS}" \
-    --output-base "$(dirname "${BASH_SOURCE[0]}")/../../../../.." \
-    --go-header-file "${scriptdir}/boilerplate.go.txt" \
+kube::codegen::gen_client \
+    --output-dir "${base_dir}/pkg/client" \
+    --output-pkg "github.com/rook/rook/pkg/client" \
+    --boilerplate "${boilerplate}" \
     --plural-exceptions "CephNFS:CephNFSes" \
+    --with-watch \
+    "${base_dir}/pkg/apis"
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:46]: Still not indented

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: build/codegen/codegen.sh (hunk 37 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: done

File: build/codegen/codegen.sh (hunk 38 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: indented now.

File: build/codegen/codegen.sh [hunk 39]
```diff
@@ -14,29 +14,33 @@
...
-    --go-header-file "${scriptdir}/boilerplate.go.txt" \
-    --plural-exceptions "CephNFS:CephNFSes" \
+source "${CODE_GENERATOR}/kube_codegen.sh"
+(
+  # because apis is a separate go module, code generation
+  # must be run from within the corresponding directory.
+  cd  ${base_dir}/pkg/apis
+
+  # run code deepcopy generation
+   kube::codegen::gen_helpers \
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:34]:
```suggestion
  kube::codegen::gen_helpers \
```

Comment (BlaineEXE) [inline build/codegen/codegen.sh:34]:
```suggestion
  kube::codegen::gen_helpers \
```

Comment (BlaineEXE) [inline build/codegen/codegen.sh:34]:
```suggestion
  kube::codegen::gen_helpers \
```

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: build/codegen/codegen.sh (hunk 39 above)
Comment (obnoxxx) [reply build/codegen/codegen.sh]: applied

Comment (obnoxxx) [reply build/codegen/codegen.sh]: done

**Review by BlaineEXE — APPROVED** [review]

File: build/codegen/codegen.sh [hunk 40]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env -S bash -e
+#!/usr/bin/env bash
```
Comment (parth-gr) [inline build/codegen/codegen.sh:1]: should we keep -e

Comment (BlaineEXE) [reply build/codegen/codegen.sh:1]: This is covered by `set -o errexit` now, later in the script
